core/src/test/java/org/apache/iceberg/TestSnapshotProducer.java

```diff
@@ -69,6 +78,95 @@ public void testManifestFileGroupSize() {
         "Must limit parallelism to avoid tiny manifests");
   }
 
+  @TestTemplate
+  void testCommitValidationFailsOnExistingWapIdInSnapshotHistory() {
+    String stagedWapId = "12345";
+    table
+        .newFastAppend()
+        .appendFile(FILE_A)
+        .set(SnapshotSummary.STAGED_WAP_ID_PROP, stagedWapId)
+        .commit();
+
+    String validationExceptionMessage =
+        String.format("Duplicate %s: %s", SnapshotSummary.STAGED_WAP_ID_PROP, stagedWapId);
+    TestingSnapshotProducer producer =
+        new TestingSnapshotProducer(table.ops())
+            .validateWith(new WapIdValidator(stagedWapId, validationExceptionMessage));
+
+    assertThatThrownBy(producer::commit)
+        .isInstanceOf(ValidationException.class)
+        .hasMessage(validationExceptionMessage);
+  }
+
+  @TestTemplate
+  void testCommitValidationStartsFromConfiguredSnapshot() {
+    String stagedWapId = "12345";
+    table
+        .newFastAppend()
+        .appendFile(TestBase.FILE_A)
+        .set(SnapshotSummary.STAGED_WAP_ID_PROP, stagedWapId)
+        .commit();
+
+    TestingSnapshotProducer producer =
+        new TestingSnapshotProducer(table.ops())
+            .validateFromSnapshot(table.currentSnapshot().snapshotId())
+            .validateWith(new WapIdValidator(stagedWapId, "empty"));
+
+    assertThatNoException().isThrownBy(producer::commit);
+    assertThat(Iterables.size(table.snapshots())).isEqualTo(2);
+  }
+
+  private static class WapIdValidator implements Consumer<Snapshot> {
+    private final String stagedWapId;
+    private final String validationErrorMessage;
+
+    private WapIdValidator(String stagedWapId, String validationErrorMessage) {
+      this.stagedWapId = stagedWapId;
+      this.validationErrorMessage = validationErrorMessage;
```

Comment:
This is a test, but I would not want people looking at this to copy the convention used here of passing an error message into the validator. I'd expect this to be embedded in the exception produced by validation. This is probably related to the issue below, where the actual ID is missing.